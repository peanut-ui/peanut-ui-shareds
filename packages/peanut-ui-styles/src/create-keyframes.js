import { keyframes } from '@emotion/css'

/**
 * Function for creating keyframes and generating them into unique animation classname.
 *
 * @param {Object} values - object css keyframes.
 * @returns {String} string css classname.
 */

export const createKeyframes = (values) => keyframes(values)
