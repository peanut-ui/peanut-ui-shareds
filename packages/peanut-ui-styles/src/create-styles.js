import { css, cx } from '@emotion/css'
import { isObject, isFunction } from '@chakra-ui/utils'
import { cssProperties, cssPseudoProperties, allCssProperties } from './utils/css-property.util'
import { isTokenVariable, toCssVariable } from './utils/css-variable.util'

const mergeCssClass = (cx, classes, prefix) => {
  return Object.keys(classes).reduce((previous, current) => {
    previous[current] = cx(classes[current], prefix ? `peanut-${prefix}-${current}` : null)

    return previous
  }, {})
}

const entriesCssObject = (entries) => {
  const result = {}

  Object.keys(entries).forEach((index) => {
    const [key, value] = entries[index]

    result[key] = value
  })

  return result
}

const resolveStyles = (cssObject) => {
  let computedStyles = {}
  let computedCssObject = cssObject

  for (let key in computedCssObject) {
    let value = computedCssObject[key]

    if (key in allCssProperties) {
      if (key in cssProperties) key = cssProperties[key]
      if (key in cssPseudoProperties) key = cssPseudoProperties[key]
      if (isObject(value)) {
        computedStyles[key] = resolveStyles(value)
      } else {
        if (isTokenVariable(value)) computedStyles[key] = toCssVariable(value, 'peanut').reference
        else computedStyles[key] = value
      }
    } else delete computedStyles[key]
  }

  return computedStyles
}

export const createStyles = (cssObject) => {
  const getCssObject = isFunction(cssObject) ? cssObject : () => cssObject

  const getStyles = (params, prefix) => {
    const cssObject = getCssObject(params)
    const classes = entriesCssObject(
      Object.keys(cssObject).map((key) => {
        const mergedStyles = cx(css(resolveStyles(cssObject[key])))

        return [key, mergedStyles]
      })
    )

    return { classes: mergeCssClass(cx, classes, prefix), cx }
  }

  return getStyles
}
