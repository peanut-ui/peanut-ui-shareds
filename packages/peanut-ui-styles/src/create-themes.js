import { isObject } from '@chakra-ui/utils'
import { toCssVariable } from './utils/css-variable.util'

export const createThemes = (tokenObject) => {
  let cssVariable = {}
  let cssMap = {}
  let referenceKeys = []

  const resolveThemes = (tokenObject) => {
    for (let [key, value] of Object.entries(tokenObject)) {
      if (isObject(value)) {
        referenceKeys.push(key)
        resolveThemes(value)
      } else {
        const keys = referenceKeys.join('.').concat('.', key)
        const { variable, reference } = toCssVariable(keys, 'peanut')

        cssVariable[variable] = value
        cssMap[keys] = { value, variable, reference }
      }
    }

    referenceKeys = []
  }

  resolveThemes(tokenObject)

  return { cssVariable, cssMap }
}
