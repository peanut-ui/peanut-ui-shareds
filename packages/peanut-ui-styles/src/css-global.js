import { injectGlobal } from '@emotion/css'

const toKebabCase = (value) => value.replace(/[A-Z]/g, (letter) => `-${letter.toLowerCase()}`)
const replaceString = (value, newValue, search) => value.replace(search, newValue)

export const cssGlobal = (cssObject) => {
  let cssMerged = ''

  for (let [key, value] of Object.entries(cssObject)) {
    let cssComputed = ''

    Object.keys(value).map((key) => (cssComputed += `${toKebabCase(key)}:${value[key]};`))
    cssMerged += replaceString(`${key}{${cssComputed}}`, '', /\s+/g)
  }

  injectGlobal`${cssMerged}`

  return cssMerged
}
