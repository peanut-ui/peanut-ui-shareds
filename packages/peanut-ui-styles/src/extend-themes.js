export const extendThemes = (oldCssVariable, newCssVariable) => {
  const cssVariable = Object.assign(oldCssVariable.cssVariable, newCssVariable.cssVariable)
  const cssMap = Object.assign(oldCssVariable.cssMap, newCssVariable.cssMap)

  return { cssVariable, cssMap }
}
