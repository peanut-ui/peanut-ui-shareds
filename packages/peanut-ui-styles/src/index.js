export * from './create-keyframes'
export * from './create-styles'
export * from './create-themes'
export * from './css-global'
