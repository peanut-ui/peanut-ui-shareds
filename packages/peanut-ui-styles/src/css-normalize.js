import { cssGlobal } from './css-global'
import { cssNormalizeStyles } from './utils/css-normalize.util'

export const cssNormalize = () => cssGlobal(cssNormalizeStyles)
