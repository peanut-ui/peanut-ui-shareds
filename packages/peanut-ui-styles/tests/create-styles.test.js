import { createStyles } from '../src'

describe('@peanut-ui/peanut-ui-styles/create-styles', () => {
  test('should convert object into classes object and cx function', () => {
    const getStyles = createStyles(() => ({
      wrapper: {
        backgroundColor: '$color.brand.100',
        paddingTop: '$space.100',
      },
      text: {
        color: '$color.brand.100',
        _hover: {
          color: '$color.brand.200',
        },
      },
    }))

    expect(getStyles()).toEqual({
      classes: {
        text: 'css-18cozed',
        wrapper: 'css-154rb2q',
      },
      cx: expect.any(Function),
    })
  })

  test('should create or check matches snapshot', () => {
    const getStyles = createStyles(() => ({
      wrapper: {
        backgroundColor: '$color.brand.100',
        paddingTop: '$space.100',
      },
      text: {
        color: '$color.brand.100',
      },
    }))

    expect(getStyles()).toMatchSnapshot()
  })
})
