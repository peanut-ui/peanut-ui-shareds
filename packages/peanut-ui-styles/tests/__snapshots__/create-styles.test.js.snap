// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@peanut-ui/peanut-ui-styles/create-styles should create or check matches snapshot 1`] = `
Object {
  "classes": Object {
    "text": "css-d090gn",
    "wrapper": "css-154rb2q",
  },
  "cx": [Function],
}
`;
